### PR TITLE
[feature](vertical_compaction) support vertical compaction

### PR DESCRIPTION
**Pull request by yixiutt**
# Proposed changes

Issue Number: close Issue #12966 

Support vertical compaction to reduce memory usage in compaction。

Mainly Design:
Split columns into several groups and record data's sequence in a RowSourceBuffer when compact key group, and use RowSourceBuffer to compact value groups. 

Result:
In clickbench test,  vertical compaction use 1/10 percent memory of  horizontal compaction and speed up 15%。

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/common/config.h [hunk 1]
```diff
@@ -246,12 +246,20 @@ CONF_Bool(enable_storage_vectorization, "true");
...
 // whether disable automatic compaction task
 CONF_mBool(disable_auto_compaction, "false");
 // whether enable vectorized compaction
 CONF_Bool(enable_vectorized_compaction, "true");
 // whether enable vectorized schema change/material-view/rollup task.
 CONF_Bool(enable_vectorized_alter_table, "true");
+// whether enable vertical compaction
+CONF_mBool(enable_vertical_compaction, "false");
+// In vertical compaction, column number for every group
+CONF_Int32(vertical_compaction_num_columns_per_group, "5");
```
Comment (dataroaring) [inline be/src/common/config.h:260]: This config can be changed online and it works according code in merger.cpp

File: be/src/olap/merger.cpp [hunk 2]
```diff
@@ -188,4 +190,137 @@ Status Merger::vmerge_rowsets(TabletSharedPtr tablet, ReaderType reader_type,
...
+    int32_t delete_sign_idx = -1;
+    // in key column compaction, seq_col real index is _block->columns() -2
+    // and delete_sign column is _block->columns() - 1
+    if (tablet_schema->keys_type() == KeysType::UNIQUE_KEYS) {
+        if (tablet_schema->has_sequence_col()) {
+            sequence_col_idx = tablet_schema->sequence_col_idx();
+            key_columns.emplace_back(sequence_col_idx);
+        }
+        delete_sign_idx = tablet_schema->field_index(DELETE_SIGN);
+        key_columns.emplace_back(delete_sign_idx);
```
Comment (dataroaring) [inline be/src/olap/merger.cpp:215]: There is not delete sign idx in some table, e.g. created in an older doris.

File: regression-test/suites/compaction/test_vertical_compaction_uniq_keys.groovy [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            UNIQUE KEY(`user_id`, `date`, `datev2`, `datetimev2_1`, `datetimev2_2`, `city`, `age`, `sex`) DISTRIBUTED BY HASH(`user_id`)
+            PROPERTIES ( "replication_num" = "1" );
+        """
+
+        sql """ INSERT INTO ${tableName} VALUES
+             (1, '2017-10-01', '2017-10-01', '2017-10-01 11:11:11.110000', '2017-10-01 11:11:11.110111', 'Beijing', 10, 1, '2020-01-01', '2020-01-01', '2017-10-01 11:11:11.170000', '2017-10-01 11:11:11.110111', '2020-01-01', 1, 30, 20)
+            """
+
+        sql """ INSERT INTO ${tableName} VALUES
+             (1, '2017-10-01', '2017-10-01', '2017-10-01 11:11:11.110000', '2017-10-01 11:11:11.110111', 'Beijing', 10, 1, '2020-01-02', '2020-01-02', '2017-10-01 11:11:11.160000', '2017-10-01 11:11:11.100111', '2020-01-02', 1, 31, 19)
```
Comment (zhannngchen) [inline regression-test/suites/compaction/test_vertical_compaction_uniq_keys.groovy:124]: don't consider delete sign and delete query?

File: be/src/vec/olap/vertical_merge_iterator.h [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    uint32_t order() const { return _order; }
+
+    void set_is_same(bool is_same) const { _is_same = is_same; }
+
+    bool is_same() { return _is_same; }
+
+    void add_cur_batch() { _cur_batch_num++; }
+
+    void reset_cur_batch() { _cur_batch_num = 0; }
```
Comment (zhannngchen) [inline be/src/vec/olap/vertical_merge_iterator.h:168]: This method is not used?

File: be/src/olap/compaction.cpp [hunk 5]
```diff
@@ -112,14 +113,50 @@ Status Compaction::quick_rowsets_compact() {
...
+    }
+
     _tablet->data_dir()->disks_compaction_score_increment(permits);
     _tablet->data_dir()->disks_compaction_num_increment(1);
     Status st = do_compaction_impl(permits);
     _tablet->data_dir()->disks_compaction_score_increment(-permits);
     _tablet->data_dir()->disks_compaction_num_increment(-1);
+
+    if (config::enable_compaction_checksum) {
+        EngineChecksumTask checksum_task(_tablet->tablet_id(), _tablet->schema_hash(),
```
Comment (dataroaring) [inline be/src/olap/compaction.cpp:131]: EngineChecksumTask does not do merge for unique key, so it does not work for now?

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 6]
```diff
@@ -36,7 +36,16 @@ BetaRowsetReader::BetaRowsetReader(BetaRowsetSharedPtr rowset)
...
+    _read_options.column_predicates.clear();
+    _read_options.col_id_to_predicates.clear();
+    _read_options.col_id_to_del_predicates.clear();
+    _read_options.key_ranges.clear();
+}
+
+Status BetaRowsetReader::get_segment_iterators(RowsetReaderContext* read_context,
+                                               std::vector<RowwiseIterator*>* out_iters) {
     RETURN_NOT_OK(_rowset->load());
     _context = read_context;
```
Comment (dataroaring) [inline be/src/olap/rowset/beta_rowset_reader.cpp:50]: This assignment is duplicated with init? BTW, get methods should not set member.

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 7]
```diff
@@ -47,68 +56,69 @@ Status BetaRowsetReader::init(RowsetReaderContext* read_context) {
...
+                &_read_options.col_id_to_del_predicates);
         // if del cond is not empty, schema may be different in multiple rowset
-        can_reuse_schema = read_options.col_id_to_del_predicates.empty();
+        _can_reuse_schema = _read_options.col_id_to_del_predicates.empty();
     }
-
-    if (!can_reuse_schema || _context->reuse_input_schema == nullptr) {
+    // In vertical compaction, every column group need new schema
+    if (read_context->is_vertical_compaction) {
+        _can_reuse_schema = false;
```
Comment (dataroaring) [inline be/src/olap/rowset/beta_rowset_reader.cpp:81]: Same as above.

File: be/src/olap/rowset/segment_v2/segment_writer.h [hunk 8]
```diff
@@ -120,11 +130,16 @@ class SegmentWriter {
...
+    std::unique_ptr<vectorized::OlapBlockDataConvertor> _olap_data_convertor;
     // used for building short key index or primary key index during vectorized write.
     std::vector<const KeyCoder*> _key_coders;
     std::vector<uint16_t> _key_index_size;
     size_t _short_key_row_pos = 0;
+
+    std::vector<uint32_t> _column_ids;
+    bool _has_key = true;
+    // written when add particial columns
+    uint32_t _num_rows_written = 0;
```
Comment (dataroaring) [inline be/src/olap/rowset/segment_v2/segment_writer.h:142]: The variable _num_rows_written is confused. It seems that num_rows_written works if users used segment writer in the order write keys, flush keys, writer colums? We should add more comment here.

// used to count rows to SegmentWriter, in vertical compaction, num_rows_written = actual_row_count * column groups number.  or we can find a better name.

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 9]
```diff
@@ -299,11 +322,12 @@ uint64_t SegmentWriter::estimate_segment_size() {
     return size;
 }
 
-Status SegmentWriter::finalize(uint64_t* segment_file_size, uint64_t* index_size) {
-    // check disk capacity
-    if (_data_dir != nullptr && _data_dir->reach_capacity_limit((int64_t)estimate_segment_size())) {
-        return Status::InternalError("disk {} exceed capacity limit.", _data_dir->path_hash());
+Status SegmentWriter::finalize_columns(uint64_t* index_size) {
+    if (_has_key) {
```
Comment (dataroaring) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:326]: Please add comment for usage. e.g. appendkey finalize_columns  or we can add a method finalize_key_columns and it takes row_count as a argument.

**Review by dataroaring — CHANGES_REQUESTED** [review]
Some minor reviews.

**Comment by morningman** [comment]
Hi @yixiutt , I think this is a breaking change to Doris core feature, so I created a new branch:
https://github.com/apache/doris/tree/compaction_opt for this feature dev.

And I have pushed the PR: opt compaction task producer and quick compaction (#13495) to it.
I will close this PR, and please push this PR to branch `compaction_opt` for testing

**Pull request closed without merging**